### PR TITLE
Support multiple execution targets

### PR DESCRIPTION
**Pull request by grzegorz8**
So far only local execution target has been supported. Now a user can run SQLs locally, remotely or on yarn.

### CODE REVIEW

File: streaming_jupyter_integrations/magics.py [hunk 1]
```diff
@@ -76,7 +67,104 @@ def __init__(self, shell: Any):
...
+        jm_hostname, jm_port = self._find_session_jm_address(rm_hostname, rm_port, yarn_application_id)
+        self._flink_connect_remote(jm_hostname, jm_port)
+
+    def _find_session_jm_address(self, rm_hostname: str, rm_port: int, yarn_application_id: str) -> Tuple[str, int]:
+        if rm_hostname and rm_port:
+            rm_client = ResourceManagerClient(rm_hostname, rm_port)
+        else:
+            rm_client = ResourceManagerClient()
+        if yarn_application_id is None:
+            yarn_application_id = self._find_yarn_application_id(rm_client)
```
Comment (swtwsk) [inline streaming_jupyter_integrations/magics.py:138]: I would wrap the parameters in the `Optional[...]` type annotation.

**Review by swtwsk — CHANGES_REQUESTED** [review]
Seems mostly ok to me. I wonder if we could test some of those changes 😄? I do know that we do not have tests for magics (mostly because it would be hard to write a unit test for something like `%%flink_execute_sql` 😅), however imo few of the "private" functions you have created might benefit from mocked-up unit testing (especially considering we are passing optional arguments, which could blow up the application if a specific mix of arguments is not properly handled).

File: streaming_jupyter_integrations/magics.py (hunk 1 above)
Comment (grzegorz8) [reply streaming_jupyter_integrations/magics.py]: done

**Comment by grzegorz8** [comment]
> Seems mostly ok to me. I wonder if we could test some of those changes smile? I do know that we do not have tests for magics (mostly because it would be hard to write a unit test for something like `%%flink_execute_sql` sweat_smile), however imo few of the "private" functions you have created might benefit from mocked-up unit testing (especially considering we are passing optional arguments, which could blow up the application if a specific mix of arguments is not properly handled).

Tests added.

File: README.md [hunk 2]
```diff
@@ -8,18 +8,109 @@
...
 %load_ext streaming_jupyter_integrations.magics
 ```
 
+Then you need to decide which _execution mode_ and _execution target_ to choose.
+
+```python
+%flink_connect --execution-mode [mode] --execution-target [target]
+```
+
+By default, the streaming execution mode and local execution target are used.
```
Comment (tnazarew) [inline README.md:33]: maybe emphasize mode and target with `this` so they are more visible

File: tests/test_yarn.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    @staticmethod
+    def _mock_cluster_apps_success(apps_list):
+        responses.add(
+            responses.GET,
+            "http://example.com:8123/ws/v1/cluster/apps",
+            json={"apps": {"app": apps_list}},
+            status=200,
+        )
+
+    def _mock_cluster_app_success(self, app):
```
Comment (niladrem) [inline tests/test_yarn.py:91]: pls add @staticmethod

**Review by niladrem — CHANGES_REQUESTED** [review]
<3

File: tests/test_yarn.py (hunk 3 above)
Comment (grzegorz8) [reply tests/test_yarn.py]: fixed

File: README.md [hunk 4]
```diff
@@ -8,18 +8,109 @@
...
+    --remote-hostname example.com \
+    --remote-port 8888
+```
+
+#### YARN session execution target
+
+Running Flink in `yarn-session` mode will connect to an existing Flink session cluster running on YARN. You may specify
+the hostname and port of the YARN Resource Manager (`--resource-manager-hostname` and `--resource-manager-port`).
+If Resource Manager address is not provided, the address is extracted from Hadoop configuration files
+(`$HADOOP_CONF_DIR/yarn-site.xml` file is searched). You can also specify YARN applicationId (`--yarn-application-id`)
```
Comment (tnazarew) [inline README.md:89]: searched for? 
also can the location be specified or do you need to have your `HADOOP_CONF_DIR` defined?

File: README.md [hunk 5]
```diff
@@ -8,18 +8,109 @@
...
+to which the notebook will connect to. If `--yarn-application-id` is not specified, the notebook will connect to
+the only YARN application running on the cluster. If there are more than one running application, the command will fail.
```
Comment (tnazarew) [inline README.md:91]: I don't like those two sentences but I don't yet know how to improve them

File: README.md (hunk 4 above)
Comment (grzegorz8) [reply README.md]: I rephrased it a bit.
I think this is a common approach for hadoop deps to expect that HADOOP_CONF_DIR is already defined.

File: README.md (hunk 2 above)
Comment (grzegorz8) [reply README.md]: done

**Review by niladrem — APPROVED** [review]

**Review by swtwsk — APPROVED** [review]

**Pull request merged**